### ISSUE TITLE
Pre-Set nginx Headers are ignored/unset for cached pages

### ISSUE THREAD

**Issue by JanThiel** [SatelliteWP/rocket-nginx#112]
**Describe the bug**
Well described here: https://www.peterbe.com/plog/be-very-careful-with-your-add_header-in-nginx

If you add any global nginx headers with add_header they are discarded by 

```
# Add header to HTML cached files
location ~ /wp-content/cache/wp-rocket/.*html$ {
        etag on;
        add_header Vary "Accept-Encoding, Cookie";
        ....
}

# Do not gzip cached files that are already gzipped
location ~ /wp-content/cache/wp-rocket/.*_gzip$ {
        etag on;
        gzip off;
        types {}
        default_type text/html;
        add_header Content-Encoding gzip;
        ....
}
``` 

**Versions**
What version of Nginx are you using ?
1.16.1

What version of Rocket-Nginx are you using ?
origin/master (2.1.1)

What version of WP Rocket are you using ?
3.3.7

Are you using Nginx as a reverse proxy (with Apache for instance) ?
No

**To Reproduce**
Steps to reproduce the behaviour:
Add any header to you vhost config before you add  include /etc/nginx/rocket-nginx/default.conf; like:
```
server {
        ....
        add_header X-XSS-Protection "1; mode=block";
        add_header X-Content-Type-Options "nosniff";
        include /etc/nginx/rocket-nginx/default.conf;
       ....
}
```

1. Call a page that is not cached: You will see the headers from the server block + the rocket-nginx headers that are NOT in the location block.

2. Call a cached page and you will see that they are missing.

Did you activate the debug in Rocket-Nginx ? Please do and include any headers.
not necessary

**Expected behavior**
The previously set global headers should remain set for cached pages. Thus no add_header must be called within rocket-nginx location blocks for cached pages. Outside of the location blocks should work fine.

**Additional context**

**Comment by maximejobin** [comment]
This is because of the way Nginx is made. That is why I have created header configurations.

You have to manually add them in the `rocket-nginx.ini` file like this:

    http_header[X-XSS-Protection] = "1; mode=block"
    http_header[X-Content-Type-Options] = "nosniff"

    css_header[X-XSS-Protection] = "1; mode=block"
    css_header[X-Content-Type-Options] = "nosniff"

    js_header[X-XSS-Protection] = "1; mode=block"
    js_header[X-Content-Type-Options] = "nosniff"

    medias_header[X-XSS-Protection] = "1; mode=block"
    medias_header[X-Content-Type-Options] = "nosniff"

**Comment by JanThiel** [comment]
Ah, ok, I got your approach. Still, this behaviour **of nginx** is highly unexpected and thus I think that an add-on solution like rocket-nginx (that is aware of this design issue) should try to avoid triggering this pitfall. Because most user will not know about this and will be caught off-guard when they acutally expect no changes in there global headers.

What about moving all the add_header commands before or after the location blocks and only work with the variables to set their value accordingly? (Like https://stackoverflow.com/a/41135845 )
It should be easy to add an additional if (-f ) file detection check to see whether there is a cached version of the current page and unset all variables if there is no expected file.

What do you think?

**Comment by JanThiel** [comment]
Hi @maximejobin I did some more testing, as I do not agree with the current approach.
rocket-nginx is an addon to existing nginx configs. It should not overrule pre-existing or globally set headers. Neither should users / admins be forced to repeat all the headers to the rocket-nginx configs. That's bad practice (Don't Repeat Yourself) and also leads to problems as you have to sync the rocket-nginx and non-rocket-nginx world.

I do not believe, that your approach setting the headers in the location blocks is necessaery at all.

Check out the attached config:
[default-test.txt](https://github.com/SatelliteWP/rocket-nginx/files/3639709/default-test.txt)

All I did was adding the values you set within the location blocks to default variables and adjusting their values. Then I moved the add_header block after your location blocks. If the variables are empty nginx will not set the header.

Adding more variables:
```
# Do not alter theses values
#
set $rocket_bypass 1;				# Should NGINX bypass WordPress and call cache file directly ?
set $rocket_encryption "";			# Is GZIP accepted by client ?
set $rocket_file "";				# Filename to use
set $rocket_is_bypassed "No";		# Header text added to check if the bypass worked or not. Header: X-Rocket-Nginx-Serving-Static
set $rocket_reason "";				# Reason why cache file was not used. If cache file is used, what file was used
set $rocket_https_prefix "";		# HTTPS prefix to use when cached files are using HTTPS
set $rocket_hsts 0;					# Is HSTS is off (0) by default. Will be turned on (1) if request is HTTPS

set $rocket_vary "Accept-Encoding, Cookie";
set $rocket_cache_control "no-cache, no-store, must-revalidate";
set $rocket_content_encoding "";

# HSTS value
set $rocket_hsts_value "";
```
Unsetting them for not cached files:
```
# Clear variables if debug is not needed
if ($rocket_debug = 0) {
	set $rocket_reason "";
	set $rocket_file "";
}

# Clear headers if bypass is disabled
if ($rocket_bypass = 0) {
	set $rocket_vary "";
	set $rocket_cache_control "";
	set $rocket_content_encoding "";
}
```

Fixing the location blocks:
```
# Add header to HTML cached files
location ~ /wp-content/cache/wp-rocket/.*html$ {
	etag on;
}

# Do not gzip cached files that are already gzipped
location ~ /wp-content/cache/wp-rocket/.*_gzip$ {
	etag on;
	gzip off;
	types {}
	default_type text/html;
}

# Setting the headers
add_header Vary $rocket_vary;
add_header Cache-Control $rocket_cache_control;
add_header Strict-Transport-Security "$rocket_hsts_value";

# Debug header (when file is not cached)
add_header X-Rocket-Nginx-Serving-Static $rocket_is_bypassed;
add_header X-Rocket-Nginx-Reason $rocket_reason;
add_header X-Rocket-Nginx-File $rocket_file;
```

This simple change allows a peaceful coexistance of global nginx headers as well as setting individual headers for the wp-rocket files.

Before:
![image](https://user-images.githubusercontent.com/3857275/65387221-5597bb80-dd45-11e9-882f-dd079e164f71.png)

After:
![image](https://user-images.githubusercontent.com/3857275/65387215-46187280-dd45-11e9-94ca-fbab8b45e2af.png)

Is there anything I missed why the headers must be set in the location block? Based on my tests, there does not seem to be a technical requirement to do so.

Have a great sunday,
Jan

**Comment by maximejobin** [comment]
We have to validate if ALL `add_headers` can be removed.

If they can, I will surely do it. If not, it's useless to half remove them. I think it was added because a specific rule needed to be added at a specific place. Previous issues to be browsed to confirm if my memory is not playing me.

**Comment by JanThiel** [comment]
Hey @maximejobin any progress here?

From what I can tell you, the "basic" `add_headers` rules which come if you do not add anything manually to the rocket-nginx.conf file do not pose any issue. Because that was the case that I tested and proved working above. Thus, I cannot imagine any case in which `add_headers`should act differently.

So there might be - although I do not believe there are - some ultra-edge cases that might be an issue. Although I do not believe that many people might be using these cases as it makes not that much sense to add special configs for the wp-rocket cached files only.

Have a great night,
Jan

**Comment by JanThiel** [comment]
To have something "real" to talk about I put everything into a PR: https://github.com/SatelliteWP/rocket-nginx/pull/116

Tested it on our live site already.

**Comment by jasperf** [comment]
Interesting debate. Just bumped into this excellent script. Also saw the pull request. Seemed to have some good improvements, but conversation here has not been finished and pull request seems to have been closed..?

**Comment by maximejobin** [comment]
I'm interested into making it better and work 'out of the box' if possible.

The pull request isn't closed at all. I think we should spin a new branch and iterate from it. There were modifications made in the original config (e.g. number of days for CSS cache). This should be made a variable.

I want to thank you @JanThiel for submitting something. Let's make version 3.0 and 'rethink' everything that makes no sense anymore. I don't think it has to be backward compatible or anything so if something doesn't make sense anymore, let's remove it.

**Comment by maximejobin** [comment]
I just don't remember exactly why we had GZIP and non-gzip headers (that you dropped, @JanThiel). Once we figure this out, it might go fast to make the script even better.

**Comment by JanThiel** [comment]
Hey @maximejobin thanks for your open mind on this :-) Let me collect some of the open issues and make it a V3. I already scanned them and some of them should be easy to address right away. Especially the clean-up and naming things.

The caching time should be a variable of course. That make sense.
If you come up with some headers that should be distinguished between GZIP / non-GZIP let's think it through when we pass that bridge. In the last 7 years we used nginx, I never had the need to handle any functional headers specific for gzip / non-gzip.

**Comment by maximejobin** [comment]
@JanThiel Make sure you check the dev branch I already started. Query strings are "almost" managed already!

It is my pleasure to open the project to you so we can make it better. That is why it is open source!

**Comment by JanThiel** [comment]
Hey @maximejobin hope you are doing well!

Happy StarWars day ;-)

As a short update for you: We did some progress on the v3 code but stumbled over some "unlucky" design decissions made by the WP-Rocket team and stoped work at that point. They seem to only focus on apache and do not seem to be willingly to really support nginx at the moment. Although the required changes in code and mindset are quite small and basic...

I do not know if you are connected with someone over at WP-Rocket, but if so, a little support in the demand for better nginx support would be highly appreciated. It should highly improve the v3 config performance. 

If they remain in a way where they treat nginx users only second class, we are sadly forced to rethink our decission to use wp-rocket on our nginx infrastructure... :-/

Find the discussion here:
https://github.com/wp-media/wp-rocket/issues/2500

If they agree to do some nginx polishing, there are several more topics I would have to discuss with them to make more features work with nginx. Or - the other way - disabling features in the WP-Rocket UI when nginx is detected. Currently some feature toggles are misleading as they simply do not work on nginx or also kind of break the page caching.

Have a great start into the new week!

Jan

**Comment by maximejobin** [comment]
This will be fixed in version 3.0 as you will be able to include files in every section.

That way, there is no repetition and it will be efficient. The "plug & play" is not possible though because of the way Nginx is built.

**Comment by maximejobin** [comment]
Fixed in version 3.0. You can now add files to include with your desired headers.
